Core/include/Acts/SpacePointFormation/detail/SpacePointBuilder.ipp

```diff
@@ -0,0 +1,673 @@
+         iMeasurementsBack < measurementsBack.size(); iMeasurementsBack++) {
+      auto [gposFront, gcovFront] =
+          globalCoords(gctx, *(measurementsFront[iMeasurementsFront]));
+      auto [gposBack, gcovBack] =
+          globalCoords(gctx, *(measurementsBack[iMeasurementsBack]));
+
+      currentDiff = detail::differenceOfMeasurementsChecked(
+          gposFront, gposBack, m_config.vertex, m_config.diffDist,
+          m_config.diffPhi2, m_config.diffTheta2);
+
+      // Store the closest Measurements (distance and index) calculated so far
+      if (currentDiff < diffMin && currentDiff >= 0.) {
+        diffMin = currentDiff;
+        measurementMinDist = iMeasurementsBack;
+      }
+    }
+
+    // Store the best (=closest) result
+    if (measurementMinDist < measurementsBack.size()) {
+      std::pair<const Measurement*, const Measurement*> measurementPair;
+      measurementPair = std::make_pair(measurementsFront[iMeasurementsFront],
+                                       measurementsBack[measurementMinDist]);
+      measurementPairs.push_back(measurementPair);
+    }
+  }
+}
+
+template <typename spacepoint_t>
+template <template <typename...> typename container_t>
+void SpacePointBuilder<spacepoint_t>::calculateDoubleHitSpacePoints(
+    const GeometryContext& gctx,
+    const std::vector<std::pair<const Measurement*, const Measurement*>>&
+        measurementPairs,
+    std::back_insert_iterator<container_t<spacepoint_t>> spacePointIt) const {
+  // Source of algorithm: Athena, SiSpacePointMakerTool::makeSCT_SpacePoint()
+
+  detail::SpacePointParameters spaPoPa;
+
+  // Walk over every found candidate pair
+  for (const auto& mp : measurementPairs) {
+    // Calculate the ends of the SDEs
+    const auto& ends1 = endsOfStrip(gctx, *(mp.first));
+    const auto& ends2 = endsOfStrip(gctx, *(mp.second));
+
+    spaPoPa.q = ends1.first - ends1.second;
+    spaPoPa.r = ends2.first - ends2.second;
+
+    const auto& slink_front = getSourceLink(*(mp.first));
+    const auto& slink_back = getSourceLink(*(mp.second));
+
+    boost::container::static_vector<const SourceLink*, 2> slinks;
+    slinks.emplace_back(slink_front);
+    slinks.emplace_back(slink_back);
+
+    double resultPerpProj;
+
+    if (m_config.usePerpProj) {  // for cosmic without vertex constraint
+      resultPerpProj = detail::calcPerpendicularProjection(
+          ends1.first, ends2.first, spaPoPa.q, spaPoPa.r);
+      if (resultPerpProj <= 0.) {
+        Vector3 pos = ends1.first + resultPerpProj * spaPoPa.q;
+        double theta = acos(spaPoPa.q.dot(spaPoPa.r) /
+                            (spaPoPa.q.norm() * spaPoPa.r.norm()));
+        const auto gcov =
+            calcGlobalVars(gctx, *(mp.first), *(mp.second), theta);
+        const double varRho = gcov[0];
+        const double varZ = gcov[1];
+
+        auto sp = spacepoint_t(pos, varRho, varZ, std::move(slinks));
+        spacePointIt = std::move(sp);
+        continue;
```

Comment:
no need to `continue` since you are already at the end of the loop

Reply:
removed